android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java

```diff
@@ -26,76 +30,257 @@
+            return Completable.error(new IOException("Failed to delete upload file: " + filename));
+        }
+    }
+
+    /**
+     * @param uploadId Bridge upload identifier
+     * @return upload validation status single
+     */
+    @NonNull
+    Single<UploadValidationStatus> getUploadValidationStatus(@NonNull String uploadId) {
+        checkNotNull(uploadId, "uploadId required");
+
+        return RxUtils.toBodySingle(api.getUploadStatus(uploadId));
+    }
+
+    @NonNull
+    Completable uploadToS3(UploadFile uploadFile, UploadSession session) {
         File file = getFile(uploadFile.filename);
+        checkArgument(file.exists(), "Non-existent file: " + file.getAbsolutePath());
+
+        Single<UploadSession> sessionSingle = Single.just(session)
+                .flatMap(uploadSession -> {
+                    DateTime desiredMinimumExpiration = DateTime.now()
+                            .plusMinutes(UPLOAD_EXPIRY_WINDOW_MINUTES);
+
+                    if (uploadSession.getExpires().isBefore(desiredMinimumExpiration)) {
+                        // get a fresh upload session
+                        return getUploadSession(uploadFile);
+                    }
+                    // reuse current upload session
+                    return Single.just(session);
+                });
 
         RequestBody requestBody = RequestBody.create(MediaType.parse(uploadFile.contentType), file);
 
-        return RxUtils.toBodySingle(
-                getS3Service(session).uploadToS3(
-                        session.getUrl(),
-                        requestBody,
-                        uploadFile.md5Hash,
-                        uploadFile.contentType))
-                .toCompletable()
-                .doOnCompleted(() -> {
-                    // TODO: update upload status in DB, delete file
-                    LOG.info("Upload succeeded for id: " + session.getId());
-                }).doOnError(t -> {
-                    LOG.info("Couldn't upload to s3", t);
-                }).andThen(
-                        RxUtils.toBodySingle(api.completeUploadSession(session.getId())
-                        ).onErrorReturn((t) -> {
-                            LOG.info("Failed to call upload complete, server will recover", t);
-                            return session;
-                        })).toCompletable()
-                .andThen(getStatus(session.getId()));
+        return sessionSingle.flatMap(freshSession -> RxUtils.toBodySingle(
+                getS3Service(freshSession)
+                        .uploadToS3(
+                                freshSession.getUrl(),
+                                requestBody,
+                                uploadFile.md5Hash,
+                                uploadFile.contentType)))
+                .doOnSuccess(aVoid -> {
+                    LOG.info("S3 upload succeeded for id: " + session.getId());
+
+                    // call upload complete on a computation thread
```

Comment:
Fixed, you're right I didn't mean to do that. TUrns out, my attempt to run it on computation thread wouldn't have worked, because RxUtils.toBodySingle already called subscribeOn(Schedulers.io())